### ISSUE TITLE
[support] Expert model parallel size allowed to exceed DP ranks.

### ISSUE THREAD

**Issue by dhineshkumar-r** [NVIDIA-NeMo/Megatron-Bridge#4980]
### User problem

Hey Folks,
Please help me understand the observation.
I'm fine-tuning(SFT) Qwen3.5-122B with 32GPUs(H200).
As the length of longest sequence in training data is ~42K, I set expert-model-parallel-size=16(EMP) and CP=16. PP=1, TP=1 and DP=2. Since 16 GPUs form a DP group, I set EMP=16 as I was under the impression that it can't exceed DP ranks. However, the job runs without issues even if I set EMP=32. 

1. How exactly are experts sharded when ranks per DP and EMP are greater than 1? For ex, in the above case 16GPUs form a DP group. Does each GPU in a DP group hold paramaeters and optimizer state of attention modules of 48 layers and 12,288 experts (16 experts per layer x 48 layer × 16 GPUs per DP)?
2. While PP and TP operates within DP copy, why EMP is allowed to exceed DP ranks?
3. When distributed optimizer flag is set, how would the optimizer state of experts distributed when they sharded across GPUs more than DP ranks?

Please let me know if I'm missing something or correct my understanding.

### Desired outcome

Any helpful answers or link would help me understand implementation detail better.

### Affected area

area:training

### Urgency / use case

Important but not blocking

**Comment by yaoyu-33** [comment]
Thanks for the detailed question. The observed behavior is expected: expert parallelism is not constrained by the dense data-parallel size because Megatron uses separate, overlapping process-group layouts for dense and routed-expert parameters.

For the reported configuration:

```text
world size = 32
TP = 1
PP = 1
CP = 16
ETP = 1
```

ETP is expert tensor parallelism. We normally keep ETP at `1`; it is only increased when an individual expert itself needs to be tensor-sharded. Therefore, for this configuration the expert mesh is determined by EP while ETP remains `1`.

The two data-parallel sizes are:

```text
dense DP  = world_size / (TP × PP × CP)
          = 32 / (1 × 1 × 16)
          = 2

expert DP = world_size / (ETP × EP × PP)
```

Therefore:

| EP | Expert DP | Routed experts per GPU per layer |
|---:|---:|---:|
| 16 | 2 | 256 / 16 = 16 |
| 32 | 1 | 256 / 32 = 8 |

A useful minimum-GPU rule is:

```text
min_gpus = PP × max(TP × CP, EP × ETP)
```

The dense `TP×CP` mesh and expert `EP×ETP` mesh share the same GPUs within a pipeline stage; the dimensions do not all multiply together.

For this configuration:

```text
EP=16: min_gpus = max(16, 16) = 16
EP=32: min_gpus = max(16, 32) = 32
```

Thus EP=32 fits exactly on 32 GPUs and results in expert DP=1.

One terminology clarification: the 16-rank group in this configuration is the context-parallel group, not the pure data-parallel group. The pure dense-DP group has size 2.

Regarding parameter placement:

- Since TP=1 and PP=1, every GPU holds the dense/attention parameters for all 48 layers. CP partitions sequence activations and attention computation; it does not partition model parameters.
- Qwen3.5-122B has 256 routed experts per layer.
- With EP=16, each GPU holds 16 routed experts per layer, or `16 × 48 = 768` local routed-expert modules. The complete expert set is reconstructed across 16 EP ranks and is replicated twice because expert DP=2.
- With EP=32, each GPU holds 8 routed experts per layer, or `8 × 48 = 384` local routed-expert modules. The complete expert set is distributed once across all 32 ranks because expert DP=1.

Therefore, a single GPU does not hold all `16 × 48 × 16 = 12,288` expert modules. That number represents the complete set across an EP=16 expert mesh, not the per-GPU allocation.

For the distributed optimizer, there are two stages of partitioning:

1. EP first partitions the routed-expert parameters themselves.
2. The distributed optimizer then shards optimizer state for each local parameter set over the corresponding expert-DP group.

Consequently:

- With EP=16 and expert DP=2, optimizer state for each local expert shard can be distributed across two expert-DP replicas.
- With EP=32 and expert DP=1, there is no additional expert-DP optimizer-state sharding. However, each GPU already owns half as many expert parameters because EP increased from 16 to 32.
- Dense parameters use the dense DP×CP process group for gradient reduction and distributed-optimizer sharding. In this topology that group has size `2 × 16 = 32`, assuming the default single distributed-optimizer instance.

So EP being larger than dense DP is valid and intentional; the two values describe different parameter layouts rather than nested dimensions of one DP group.

**Comment by meatybobby** [comment]
@dhineshkumar-r Are your questions answered? If you don't have further question, we can close the issue.

**Comment by dhineshkumar-r** [comment]
Hey @yaoyu-33 ,

Thanks for the comprehensive answer and examples.
The main takeaway for me is that there are two independent meshes, one for dense and expert parameters, sharing the same set of GPUs.

One followup question I have is that while it is technically possible to train with EDP != DP, is it recommended? Wouldn't DP > EDP increase communication between GPUs and hence affect GPU utilization?

Appreciate the help.
